File: RLBridge/communication.lua
--- RLBridge communication module
--- Handles dual pipe communication with persistent handles between the game and external AI system

local COMM = {}
local utils = require("utils")
local json = require("dkjson")

-- Dual pipe communication settings with persistent handles
local comm_enabled = false
local request_pipe = "/tmp/balatro_request"
local response_pipe = "/tmp/balatro_response"
local request_handle = nil
local response_handle = nil

--- Initialize dual pipe communication with persistent handles
--- Sets up persistent pipe handles with external AI system
--- @return nil
function COMM.init()
    utils.log_comm("Initializing dual pipe communication with persistent handles...")
    utils.log_comm("Note: Pipes will be opened when first needed (lazy initialization)")
    comm_enabled = true -- Enable communication, pipes will open on first use
end

--- Lazy initialization of pipe handles when first needed
--- @return boolean True if pipes are ready, false otherwise
function COMM.ensure_pipes_open()
    if request_handle and response_handle then
        return true -- Already open
    end

    -- Try to open pipes (this will block until AI creates them)
    utils.log_comm("Opening persistent pipe handles...")

    -- Open response pipe for reading (keep open)
    response_handle = io.open(response_pipe, "r")
    if not response_handle then
        utils.log_comm("ERROR: Cannot open response pipe for reading: " .. response_pipe)
        return false
    end

    -- Open request pipe for writing (keep open)
    request_handle = io.open(request_pipe, "w")
    if not request_handle then
        utils.log_comm("ERROR: Cannot open request pipe for writing: " .. request_pipe)
        if response_handle then
            response_handle:close()
            response_handle = nil
        end
        return false
    end

    utils.log_comm("Persistent pipe handles opened successfully")
    utils.log_comm("Request pipe (write): " .. request_pipe)
    utils.log_comm("Response pipe (read): " .. response_pipe)
    return true
end

--- Send game turn request to AI and get action via persistent pipe handles
--- @param game_state table Current game state data
--- @param available_actions table Available actions list
--- @return table|nil Action response from AI, nil if error
function COMM.request_action(game_state, available_actions, retry_count)
    if not comm_enabled then
        utils.log_comm("ERROR: Communication not enabled")
        return nil
    end

    -- Lazy initialization - open pipes when first needed
    if not COMM.ensure_pipes_open() then
        utils.log_comm("ERROR: Failed to open pipe handles")
        return nil
    end

    local request = {
        game_state = game_state,
        available_actions = available_actions or {},
        retry_count = retry_count,
    }

    utils.log_comm(utils.get_timestamp() .. "Sending action request for state: " ..
        tostring(game_state.state) .. " (" .. utils.get_state_name(game_state.state) .. ")")

    -- Encode request as JSON
    local json_data = json.encode(request)
    if not json_data then
        utils.log_comm("ERROR: Failed to encode request JSON")
        return nil
    end

    -- Write request to persistent handle
    request_handle:write(json_data .. "\n")
    request_handle:flush() -- Ensure data is sent immediately
    utils.log_comm(utils.get_timestamp() .. "Request sent...")

    -- Read response from persistent handle
    utils.log_comm(utils.get_timestamp() .. "about to read the respones")
    local response_json = response_handle:read("*line")

    if not response_json or response_json == "" then
        utils.log_comm("ERROR: No response received from AI")
        return nil
    end

    local response_data = json.decode(response_json)
    if not response_data then
        utils.log_comm("ERROR: Failed to decode response JSON")
        return nil
    end

    utils.log_comm("AI action: " .. tostring(response_data.action))
    return response_data
end

--- Check if pipe communication is enabled
--- Returns the current communication status
--- @return boolean True if enabled, false otherwise
function COMM.is_connected()
    return comm_enabled
end

--- Close communication
--- Terminates the persistent pipe handles with the AI system
--- @return nil
function COMM.close()
    comm_enabled = false

    if request_handle then
        request_handle:close()
        request_handle = nil
    end

    if response_handle then
        response_handle:close()
        response_handle = nil
    end

    utils.log_comm("Persistent pipe communication closed")
end

return COMM

File: ai/utils/communication.py
"""
Balatro Dual Pipe Communication Abstraction

Pure dual pipe I/O layer with persistent handles for communicating with Balatro mod.
Handles low-level pipe operations without any game logic or AI logic.

This abstraction can be used by:
- balatro_env.py (for SB3 training)
- file_watcher.py (for testing)
- Any other component that needs to talk to Balatro
"""

import json
import logging
import os
from typing import Dict, Any, Optional


class BalatroPipeIO:
    """
    Clean abstraction for dual pipe communication with Balatro mod using persistent handles
    
    Responsibilities:
    - Read JSON requests from Balatro mod via request pipe
    - Write JSON responses back to Balatro mod via response pipe
    - Keep pipe handles open persistently to avoid deadlocks
    - Provide simple, clean interface for higher-level code
    """
    
    def __init__(self, request_pipe: str = "/tmp/balatro_request", response_pipe: str = "/tmp/balatro_response"):
        self.request_pipe = request_pipe
        self.response_pipe = response_pipe
        self.logger = logging.getLogger(__name__)
        
        # Persistent pipe handles
        self.request_handle = None
        self.response_handle = None
        
        # Create pipes and open persistent handles
        self.create_pipes()
        self.open_persistent_handles()
    
    def create_pipes(self) -> None:
        """
        Create dual named pipes for communication
        
        Creates both request and response pipes if they don't exist.
        Safe to call multiple times.
        """
        for pipe_path in [self.request_pipe, self.response_pipe]:
            try:
                # Remove existing pipe if it exists
                if os.path.exists(pipe_path):
                    os.unlink(pipe_path)
                    self.logger.debug(f"Removed existing pipe: {pipe_path}")
                
                # Create named pipe
                os.mkfifo(pipe_path)
                self.logger.info(f"Created pipe: {pipe_path}")
                
            except Exception as e:
                self.logger.error(f"Failed to create pipe {pipe_path}: {e}")
                raise RuntimeError(f"Could not create communication pipe: {pipe_path}")
        
    def open_persistent_handles(self) -> None:
        """
        Open persistent handles for reading and writing
        
        Opens request pipe for reading and response pipe for writing.
        Keeps handles open to avoid deadlocks.
        """
        import time
        
        try:
            self.logger.info(f"🔧 Waiting for Balatro to connect...")
            self.logger.info(f"   Press 'R' in Balatro now to activate RL training!")
            
            # Open request pipe for reading (Balatro writes to this)
            self.request_handle = open(self.request_pipe, 'r')
            
            # Open response pipe for writing (AI writes to this)
            self.response_handle = open(self.response_pipe, 'w')
            
        except Exception as e:
            self.logger.error(f"Failed to open persistent handles: {e}")
            self.cleanup_handles()
            raise RuntimeError(f"Could not open persistent pipe handles: {e}")
    
    def wait_for_request(self) -> Optional[Dict[str, Any]]:
        """
        Wait for new request from Balatro mod using persistent handle
        
        Blocks until Balatro writes a request to the pipe.
        No timeout needed - pipes block until data arrives.
        
        Returns:
            Parsed JSON request data, or None if error
        """
        if not self.request_handle:
            self.logger.error("Request handle not open")
            return None
            
        try:
            # Read from persistent request handle
            request_line = self.request_handle.readline().strip()
            if not request_line:
                return None
            
            request_data = json.loads(request_line)
            self.logger.debug(f"📥 RECEIVED REQUEST: {request_line}")
            return request_data
            
        except json.JSONDecodeError as e:
            self.logger.error(f"Invalid JSON in request: {e}")
            return None
        except Exception as e:
            self.logger.error(f"Error reading from request pipe: {e}")
            return None
    
    def send_response(self, response_data: Dict[str, Any]) -> bool:
        """
        Send response back to Balatro mod using persistent handle
        
        Writes response data to response pipe for Balatro to read.
        
        Args:
            response_data: Response dictionary to send
            
        Returns:
            True if successful, False if error
        """
        if not self.response_handle:
            self.logger.error("Response handle not open")
            return False
            
        try:
            # Write to persistent response handle
            json.dump(response_data, self.response_handle)
            self.response_handle.write('\n')  # Important: newline for pipe communication
            self.response_handle.flush()  # Force write to pipe immediately
            
            self.logger.debug(f"📤 SENT RESPONSE: {json.dumps(response_data)}")
            return True
            
        except Exception as e:
            self.logger.error(f"❌ ERROR sending response: {e}")
            import traceback
            self.logger.error(f"❌ TRACEBACK: {traceback.format_exc()}")
            return False
    
    def cleanup_handles(self):
        """
        Close persistent pipe handles
        
        Closes the open file handles.
        """
        if self.request_handle:
            try:
                self.request_handle.close()
                self.logger.debug("Closed request handle")
            except Exception as e:
                self.logger.warning(f"Failed to close request handle: {e}")
            self.request_handle = None
            
        if self.response_handle:
            try:
                self.response_handle.close()
                self.logger.debug("Closed response handle")
            except Exception as e:
                self.logger.warning(f"Failed to close response handle: {e}")
            self.response_handle = None
    
    def cleanup(self):
        """
        Clean up communication pipes and handles
        
        Closes handles and removes pipe files from the filesystem.
        """
        # Close handles first
        self.cleanup_handles()
        
        # Remove pipe files
        for pipe_path in [self.request_pipe, self.response_pipe]:
            try:
                if os.path.exists(pipe_path):
                    os.unlink(pipe_path)
                    self.logger.debug(f"Removed pipe: {pipe_path}")
            except Exception as e:
                self.logger.warning(f"Failed to remove pipe {pipe_path}: {e}")
        
        self.logger.debug("Dual pipe communication cleanup complete")
    
